### PR TITLE
Fix Incorrect Transport Response Handler Type

### LINKED ISSUE #38226: testAbortedSnapshotDuringInitDoesNotStart fails with ClassCastException
This test is failing with ClassCastException.

```
  1> org.elasticsearch.transport.RemoteTransportException: [node_sd3][127.0.0.1:34425][internal:cluster/snapshot/update_snapshot_status]
  1> Caused by: org.elasticsearch.transport.ResponseHandlerFailureTransportException: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
  1> Caused by: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
  1> 	at org.elasticsearch.snapshots.SnapshotShardsServic
```

I can't reproduce this locally but this test failed 4 times today.

```
 ./gradlew :server:integTest \
  -Dtests.seed=F8B30D4F31180D3F \
  -Dtests.class=org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT \
  -Dtests.method="testAbortedSnapshotDuringInitDoesNotStart" \
  -Dtests.security.manager=true \
  -Dtests.locale=cs \
  -Dtests.timezone=AET \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

CI:
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1768/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1763/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=amazon/217/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+internalClusterTest/262/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+default-distro-bwc-tests/326/consoleText

### LINKED ISSUE #38256: [CI] DedicatedClusterSnapshotRestoreIT#testRestoreShrinkIndex fails
Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+internalClusterTest/349/console

Using this reproduce line repeatedly, the test usually fails after a few tries locally (not every time though, but quite soon when running in a loop)
```
./gradlew :server:integTest \
  -Dtests.seed=B2F3F6260A9D92DB \
  -Dtests.class=org.elasticsearch.snapshots.DedicatedClusterSnapshotRestoreIT \
  -Dtests.method="testRestoreShrinkIndex" \
  -Dtests.security.manager=true \
  -Dtests.locale=sr-Latn-ME \
  -Dtests.timezone=America/Guadeloupe \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

The logs show several ClassCastExceptions like this:
```
12:58:03   1> [2019-02-02T07:55:51,306][WARN ][o.e.s.SnapshotShardsService] [node_t1] [test-repo:test-snap-1/wuZ3MOnpQw6EbCxig7umvw] [ShardSnapshotStatus[state=SUCCESS, nodeId=GC6yybZDQsyUP9FCZGZGkw, reason=null]] failed to update snapshot state
12:58:03   1> org.elasticsearch.transport.RemoteTransportException: [node_t1][127.0.0.1:42747][internal:cluster/snapshot/update_snapshot_status]
12:58:03   1> Caused by: org.elasticsearch.transport.ResponseHandlerFailureTransportException: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
12:58:03   1> Caused by: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
12:58:03   1> 	at org.elasticsearch.snapshots.SnapshotShardsService$3.handleResponse(SnapshotShardsService.java:511) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1108) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1189) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1169) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:54) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:47) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:30) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.onResponse(TransportMasterNodeAction.java:191) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.onResponse(TransportMasterNodeAction.java:188) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.snapshots.SnapshotShardsService$4.clusterStateProcessed(SnapshotShardsService.java:546) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.clusterStateProcessed(MasterService.java:520) ~[main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.lambda$processedDifferentClusterState$1(MasterService.java:407) ~[main/:?]
12:58:03   1> 	at java.util.ArrayList.forEach(ArrayList.java:1257) [?:1.8.0_202]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.processedDifferentClusterState(MasterService.java:407) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService.onPublicationSuccess(MasterService.java:264) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService.publish(MasterService.java:257) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:238) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:142) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [main/:?]
12:58:03   1> 	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [main/:?]
12:58:03   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [main/:?]
12:58:03   1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [main/:?]
12:58:03   1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [main/:?]
12:58:03   1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_202]
12:58:03   1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_202]
12:58:03   1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
```

Followed by a  MasterNotDiscoveredException, possibly as a consequence of the former. 
```
> Throwable #1: MasterNotDiscoveredException[null]
   >    at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$4.onTimeout(TransportMasterNodeAction.java:259)
   >    at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:322)
   >    at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:249)
   >    at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:549)
   >    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681)
   >    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   >    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   >    at java.lang.Thread.run(Thread.java:748)Throwable #2: MasterNotDiscoveredException[null]
   >    at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$4.onTimeout(TransportMasterNodeAction.java:259)
   >    at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:322)
   >    at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:249)
   >    at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:549)
   >    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681)
   >    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   >    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   >    at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by original-brownbear**
* The response type here is not empty and was always wrong but this only became visible now that 0a604e3b2496777168bb541815a8526e463ee15f was introduced
   * As a result of 0a604e3b2496777168bb541815a8526e463ee15f we started actually handling the response
of this request and logging/handling exceptions before that we simply dropped the class cast exception here quietly using the empty response handler
* Closes #38226 
* Closes #38256

----------

* Marked as `>non-issue` because this never made it into a release

### CODE REVIEW

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @original-brownbear.

**Comment by original-brownbear** [comment]
@dnhatn thanks for the quick review, bad news though:
Seems that error was just a symptom (well it's a bug too) and the test failed again in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/7125/consoleText

Looking into that now. I'd suggest if I can't fix it quickly I re-mute the tests and at least merge this fix to cut back on the noise? It seems like this could hit other tests that haven't yet failed too randomly.

**Comment by original-brownbear** [comment]
Fixed in 7b26929

The issue was that the changed busy assert block died on a normal exception when concurrently loading the repository data (I guess we could look into making the behaviour of the FS repository here nicer eventually):

```
ERROR   0.48s J3 | SharedClusterSnapshotRestoreIT.testAbortedSnapshotDuringInitDoesNotStart <<< FAILURES!
   > Throwable #1: UncategorizedExecutionException[Failed execution]; nested: NoSuchFileException[/var/lib/jenkins/workspace/elastic+elasticsearch+pull-request-1/server/build/testrun/integTest/J3/temp/org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT_DF78BC25D4400645-001/tempDir-002/repos/sYzVZglzXR/indices/qCbBefHbTamRgzueAMrRvg/meta-saB1MnWJSuaATpK4KqAAVQ.dat];
   > 	at __randomizedtesting.SeedInfo.seed([DF78BC25D4400645:A7A575D3E8E33D4B]:0)
   > 	at org.elasticsearch.common.util.concurrent.FutureUtils.rethrowExecutionException(FutureUtils.java:97)
   > 	at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:62)
   > 	at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:34)
   > 	at org.elasticsearch.action.ActionRequestBuilder.get(ActionRequestBuilder.java:52)
   > 	at org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT.lambda$testAbortedSnapshotDuringInitDoesNotStart$24(SharedClusterSnapshotRestoreIT.java:3686)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:846)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:832)
   > 	at org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT.testAbortedSnapshotDuringInitDoesNotStart(SharedClusterSnapshotRestoreIT.java:3685)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.nio.file.NoSuchFileException: /var/lib/jenkins/workspace/elastic+elasticsearch+pull-request-1/server/build/testrun/integTest/J3/temp/org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT_DF78BC25D4400645-001/tempDir-002/repos/sYzVZglzXR/indices/qCbBefHbTamRgzueAMrRvg/meta-saB1MnWJSuaATpK4KqAAVQ.dat
   > 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
   > 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
   > 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
   > 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
   > 	at java.nio.file.Files.newByteChannel(Files.java:361)
   > 	at java.nio.file.Files.newByteChannel(Files.java:407)
   > 	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
   > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
   > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
   > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
   > 	at org.apache.lucene.mockfile.HandleTrackingFS.newInputStream(HandleTrackingFS.java:92)
   > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
   > 	at org.apache.lucene.mockfile.HandleTrackingFS.newInputStream(HandleTrackingFS.java:92)
   > 	at java.nio.file.Files.newInputStream(Files.java:152)
   > 	at org.elasticsearch.common.blobstore.fs.FsBlobContainer.readBlob(FsBlobContainer.java:120)
   > 	at org.elasticsearch.snapshots.mockstore.BlobContainerWrapper.readBlob(BlobContainerWrapper.java:48)
   > 	at org.elasticsearch.snapshots.mockstore.MockRepository$MockBlobStore$MockBlobContainer.readBlob(MockRepository.java:318)
   > 	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.readBlob(ChecksumBlobStoreFormat.java:101)
   > 	at org.elasticsearch.repositories.blobstore.BlobStoreFormat.read(BlobStoreFormat.java:90)
   > 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getSnapshotIndexMetaData(BlobStoreRepository.java:576)
   > 	at org.elasticsearch.snapshots.SnapshotsService.snapshotShards(SnapshotsService.java:635)
   > 	at org.elasticsearch.action.admin.cluster.snapshots.status.TransportSnapshotsStatusAction.buildResponse(TransportSnapshotsStatusAction.java:236)
   > 	at org.elasticsearch.action.admin.cluster.snapshots.status.TransportSnapshotsStatusAction.masterOperation(TransportSnapshotsStatusAction.java:99)
   > 	at org.elasticsearch.action.admin.cluster.snapshots.status.TransportSnapshotsStatusAction.masterOperation(TransportSnapshotsStatusAction.java:60)
   > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:127)
   > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.doRun(TransportMasterNodeAction.java:208)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```

**Comment by original-brownbear** [comment]
@dnhatn can you take another look please? :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

File: server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -3684,8 +3683,13 @@ public void testAbortedSnapshotDuringInitDoesNotStart() throws Exception {
...
             // The deletion must set the snapshot in the ABORTED state
             assertBusy(() -> {
-                SnapshotsStatusResponse status = client.admin().cluster().prepareSnapshotStatus("repository").setSnapshots("snap").get();
-                assertThat(status.getSnapshots().iterator().next().getState(), equalTo(State.ABORTED));
+                try {
+                    SnapshotsStatusResponse status =
+                        client.admin().cluster().prepareSnapshotStatus("repository").setSnapshots("snap").get();
+                    assertThat(status.getSnapshots().iterator().next().getState(), equalTo(State.ABORTED));
+                } catch (Exception e) {
+                    throw new AssertionError(e);
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: maybe add a comment to say that here we rethrow in AssertionError to force assertBusy to retry.

**Review by dnhatn — APPROVED** [review]
Thanks for an extra iteration @original-brownbear.

File: server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: Sure, I added it :)

Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: @DaveCTurner thanks! I think I have an idea where this is coming from still ... Will fix tomorrow :)

**Pull request merged**